mat64/matrix.go

```diff
@@ -781,6 +767,87 @@ func MaybeFloat(fn func() float64) (f float64, err error) {
 	return fn(), nil
 }
 
+// Norm returns the specified (induced) norm of the matrix a. See
+// https://en.wikipedia.org/wiki/Matrix_norm for the definition of an induced norm.
+//
+// Valid norms are:
+//    1 - The maximum absolute column sum
+//    2 - Frobenius norm, the square root of the sum of the squares of the elements.
+//  Inf - The maximum row sum.
+// Norm will panic with ErrNormOrder if an illegal norm order is specified.
+func Norm(a Matrix, norm float64) float64 {
+	var n lapack.MatrixNorm
```

Comment:
The problem with that relative to my current code is that the work slice only needs to be allocated in certain scenarios 1-norm for dense, but 1 or inf norm for symmetric and triangular
> On Sep 24, 2015, at 12:40 AM, Dan Kortschak <notifications@github.com> wrote:
> 
> In mat64/matrix.go <https://github.com/gonum/matrix/pull/229#discussion_r40289410>:
> 
> > @@ -781,6 +767,87 @@ func MaybeFloat(fn func() float64) (f float64, err error) {
> >  	return fn(), nil
> >  }
> >  
> > +// Norm returns the specified (induced) norm of the matrix a. See
> > +// https://en.wikipedia.org/wiki/Matrix_norm for the definition of an induced norm.
> > +//
> > +// Valid norms are:
> > +//    1 - The maximum absolute column sum
> > +//    2 - Frobenius norm, the square root of the sum of the squares of the elements.
> > +//  Inf - The maximum row sum.
> > +// Norm will panic with ErrNormOrder if an illegal norm order is specified.
> > +func Norm(a Matrix, norm float64) float64 {
> > +	var n lapack.MatrixNorm
> Try this:
> 
> func Norm(a Matrix, norm float64) float64 {
>     aMat, aTrans := untranspose(a)
>     switch aMat := aMat.(type) {
>     case RawMatrixer:
>         n, work := normWork(a, norm, aTrans)
>         return lapack64.Lange(n, rma.RawMatrix(), work)
>     case RawTriangular:
>         n, work := normWork(a, norm, aTrans)
>         return lapack64.Lantr(n, rma.RawTriangular(), work)
>     case RawSymmetricer:
>         n, work := normWork(a, norm, aTrans)
>         return lapack64.Lansy(n, rma.RawSymmetric(), work)
>     }
> 
>     switch norm {
>     default:
>         panic("unreachable")
>     case 1:
>         var max float64
>         for j := 0; j < c; j++ {
>             var sum float64
>             for i := 0; i < r; i++ {
>                 sum += math.Abs(a.At(i, j))
>             }
>             if sum > max {
>                 max = sum
>             }
>         }
>         return max
>     case 2:
>         var sum float64
>         for i := 0; i < r; i++ {
>             for j := 0; j < c; j++ {
>                 v := a.At(i, j)
>                 sum += v * v
>             }
>         }
>         return math.Sqrt(sum)
>     case math.Inf(1):
>         var max float64
>         for i := 0; i < r; i++ {
>             var sum float64
>             for j := 0; j < c; j++ {
>                 sum += math.Abs(a.At(i, j))
>             }
>             if sum > max {
>                 max = sum
>             }
>         }
>         return max
>     }
> }
> 
> func normWork(a Matrix, norm float64, trans bool) (n lapack.MatrixNorm, work []float64) {
>     switch norm {
>     case 1:
>         n = lapack.MaxColumnSum
>     case 2:
>         n = lapack.NormFrob
>     case math.Inf(1):
>         n = lapack.MaxRowSum
>     default:
>         panic(ErrNormOrder)
>     }
>     r, c := a.Dims()
>     if trans {
>         work = make([]float64, r)
>         if norm == 1 {
>             n = lapack.MaxRowSum
>         } else if norm == math.Inf(1) {
>             n = lapack.MaxColumnSum
>         }
>     } else {
>         work = make([]float64, c)
>     }
>     return n, work
> }
> —
> Reply to this email directly or view it on GitHub <https://github.com/gonum/matrix/pull/229/files#r40289410>.
>